snuba/web/db_query.py

```diff
@@ -717,7 +717,12 @@ def raw_query(
             with configure_scope() as scope:
                 if isinstance(cause, ClickhouseError):
                     error_code = cause.code
-                    scope.fingerprint = ["{{default}}", str(cause.code)]
+                    scope.fingerprint = [
+                        "{{default}}",
+                        str(cause.code),
+                        query_metadata.dataset,
+                        query_metadata.request.referrer,
```

Comment:
Why do we want to break down by referrer ? 
This will create a very large amount of groups every time we have a system issue. I would advise against.

Reply:
@fpacifici i discussed this with @evanh, but my understanding is that the current clickhouse error is hard to monitor because it conflates all the datasets

"referrer" is imo a good approximation of "where did this issue come from", a piece of information typically provided by a stacktrace

there are not that many _actually observed_ `(dataset, referrer)` combinations: https://sentry.io/organizations/sentry/issues/3111024867/events/9498896d977f4c2abc70a0edc9c705da/tags/transaction/?project=300688

Reply:
I think this addresses the more common case, which is intermittent Clickhouse errors. It is true that a system wide problem would create many issues, but that is a rarer case. Also this applies only to errors  with a Clickhouse code, which is a smaller slice of potential problems.

Reply:
what is the expected cardinality of referrer? if it is less than the number of possible stacktraces, our system really should be able to handle it...

Reply:
>Also this applies only to errors with a Clickhouse code, which is a smaller slice of potential problems.

I don't see any branching specific to existence of code, I think all clickhouse errors have one, no?

Reply:
I don't think this is a rare case.

The most common error we generate is the project rate limitier error. The number of issues will multiply. by the number of referrer (hundreds). 
I would still strongly suggest using a tag for this and not a new issue.

Reply:
> I think this addresses the more common case, which is intermittent Clickhouse errors. It is true that a system wide problem would create many issues, but that is a rarer case.

Not following. While the larger issues are less common, having them properly grouped allows a proper investigation  when they happen. If we generate hundreds of issues because a clickhouse node decided it was receiving too many concurrent queries or stopped responding on the network, even counting the occurrences will be very hard. 

Again, this seems the right problem to be solved with a tag, which provides the same querying ability without breaking grouping.

Reply:
The problem we are trying to solve is the day to day work of triaging and fixing bugs in the Sentry dashboard. A tag will not help that problem. This change will make 99% of the investigations in Sentry easier. I don't know why we would optimize for the 1%, especially because I am skeptical this change will actually _hinder_ an investigation into a catastrophic incident.

Reply:
> I don't see any branching specific to existence of code

I was referring to the implicit branching. We have many errors that happen before sending the query to Clickhouse, so there is no error code.

Reply:
is it possible to determine the clickhouse error codes that indicate systematic failures and not problems that pertain to a specific query?

Reply:
We can (and have) isolated some of the error codes in this way, but the problem is that some error codes apply to both cases. For example `TIMEOUT_EXCEEDED = 159` can mean a single query is taking too long, but if it is happening across many queries at the same time it can mean Clickhouse itself is in a bad state.

Reply:
> Maybe a better compromise would be to pick the few error codes that are signals of general Clickhouse degradation (timeouts etc.) and not split those.

As discussed offline this still has to be done.